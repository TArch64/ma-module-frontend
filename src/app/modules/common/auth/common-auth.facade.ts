import {Injectable} from "@angular/core";
import {CommonAuthService, User} from "@common/auth";
import {Observable} from "rxjs";

export interface ICommonAuthFacade {
    isSignedOut: boolean;
    isSignedIn: boolean;
    authToken: string | null;
    currentUser: User | null;

    signOut(): void;
    fetchCurrentUser(): Observable<User>;
}

@Injectable()
export class CommonAuthFacade implements ICommonAuthFacade {
    constructor(
        private readonly authService: CommonAuthService,
    ) {}

    get isSignedOut(): boolean {
        return !this.isSignedIn;
    }

    get isSignedIn(): boolean {
        return this.authService.isSignedIn;
    }

    get authToken(): string | null {
        return this.authService.authToken;
    }

    get currentUser(): User | null {
        return this.authService.currentUserSnapshot;
    }

    signOut(): void {
        this.authService.signOut();
    }

    fetchCurrentUser(): Observable<User> {
        return this.authService.fetchCurrentUser();
    }
}
