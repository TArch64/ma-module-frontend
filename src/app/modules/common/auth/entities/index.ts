export { User, IUserJSON } from './user';
export { InsecureUser, IInsecureUserJSON } from './insecure-user';
