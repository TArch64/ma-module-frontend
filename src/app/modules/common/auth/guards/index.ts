export {AuthOnlyGuard} from './auth-only.guard';
export {InauthOnlyGuard} from './inauth-only.guard';
