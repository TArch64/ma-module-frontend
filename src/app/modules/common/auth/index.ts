export {CommonAuthModule} from './common-auth.module';
export {CommonAuthService, ICommonAuthService} from './services';
export {AuthOnlyGuard, InauthOnlyGuard, RoleAccessGuard} from './guards';
export {AuthInterceptor} from './interceptors';
export {UserRoles} from './enums';
export {User, IUserJSON, InsecureUser, IInsecureUserJSON} from './entities';
