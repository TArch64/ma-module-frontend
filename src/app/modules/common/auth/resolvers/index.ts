export {ActualizeCurrentUserResolver} from './actualize-current-user.resolver';
