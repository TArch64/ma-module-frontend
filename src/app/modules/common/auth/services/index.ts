export {CommonAuthService} from './common-auth.service';
