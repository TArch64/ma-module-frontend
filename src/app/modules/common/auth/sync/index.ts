export { CommonAuthSyncService } from './common-auth-sync.service';
