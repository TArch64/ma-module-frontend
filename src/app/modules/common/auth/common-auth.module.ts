import {NgModule} from "@angular/core";
import {CommonModule} from "@angular/common";
import {HttpClientModule} from "@angular/common/http";
import {CommonAuthSyncService} from "./sync";
import {AuthOnlyGuard, InauthOnlyGuard, RoleAccessGuard} from "./guards";
import {AuthInterceptor} from "./interceptors";

@NgModule({
    imports: [
        CommonModule,
        HttpClientModule
    ],
    providers: [
        CommonAuthSyncService,
        AuthOnlyGuard,
        InauthOnlyGuard,
        AuthInterceptor,
        RoleAccessGuard
    ]
})
export class CommonAuthModule {}
