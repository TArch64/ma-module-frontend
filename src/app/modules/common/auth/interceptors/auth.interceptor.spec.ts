import {Observable, of, throwError} from "rxjs";
import {HttpErrorResponse, HttpEvent, HttpEventType, HttpHandler, HttpRequest} from "@angular/common/http";
import {User, UserRoles} from "@common/auth";
import {ICommonAuthFacade} from "../common-auth.facade";
import {AuthInterceptor} from './auth.interceptor';

class MockCommonAuthFacade implements ICommonAuthFacade {
    static create(): MockCommonAuthFacade {
        return new MockCommonAuthFacade();
    }

    mockIsSignedIn = jest.spyOn<ICommonAuthFacade, 'isSignedIn'>(this, 'isSignedIn', 'get');
    mockSignOut = jest.spyOn<ICommonAuthFacade, 'signOut'>(this, 'signOut');

    isSignedOut: boolean = false;
    authToken: string | null = 'test';
    currentUser: User | null = null;

    get isSignedIn(): boolean {
        return false;
    };

    signOut(): void {}

    fetchCurrentUser(): Observable<User> {
        return of(User.fromJSON({
            id: 'test',
            role: UserRoles.ADMIN,
            username: 'test'
        }))
    }
}

class MockHttpHandler extends HttpHandler {
    static create(): MockHttpHandler {
        return new MockHttpHandler();
    }

    mockHandle = jest.spyOn<HttpHandler, 'handle'>(this, 'handle');

    handle(req: HttpRequest<any>): Observable<HttpEvent<any>> {
        return of({type: HttpEventType.Sent});
    }
}

function createInterceptor(facade: ICommonAuthFacade): AuthInterceptor {
    return new AuthInterceptor(facade);
}

function createRequest(): HttpRequest<any> {
    return new HttpRequest('GET', 'test');
}

describe('insert auth token', () => {
    test('should insert token if signed in', () => {
        const facade = MockCommonAuthFacade.create();
        facade.mockIsSignedIn.mockReturnValue(true);

        const interceptor = createInterceptor(facade);
        const handler = MockHttpHandler.create();

        interceptor.intercept(createRequest(), handler);

        expect(handler.mockHandle).toHaveBeenCalled();
        expect(handler.mockHandle.mock.calls[0][0].headers.get('Authorization')).toMatchSnapshot();
    });

    test('should not insert token if not signed in', () => {
        const facade = MockCommonAuthFacade.create();
        facade.mockIsSignedIn.mockReturnValue(false);

        const interceptor = createInterceptor(facade);
        const handler = MockHttpHandler.create();

        interceptor.intercept(createRequest(), handler);

        expect(handler.mockHandle).toHaveBeenCalled();
        expect(handler.mockHandle.mock.calls[0][0].headers.get('Authorization')).toBeFalsy();
    });
});

describe('error handler', () => {
    test('should sign out on unauthorized', async () => {
        const facade = MockCommonAuthFacade.create();
        const interceptor = createInterceptor(facade);
        const handler = MockHttpHandler.create();

        handler.mockHandle.mockReturnValue(throwError(new HttpErrorResponse({ status: 403 })))
        await interceptor.intercept(createRequest(), handler).toPromise().catch(() => {});

        expect(facade.mockSignOut).toHaveBeenCalled();
    });

    test('should rethrow on other error', async () => {
        const facade = MockCommonAuthFacade.create();
        const interceptor = createInterceptor(facade);
        const handler = MockHttpHandler.create();

        const error = new Error('test');
        handler.mockHandle.mockReturnValue(throwError(error))

        await expect(interceptor.intercept(createRequest(), handler).toPromise()).rejects.toEqual(error);
        expect(facade.mockSignOut).not.toHaveBeenCalled();
    });
});
