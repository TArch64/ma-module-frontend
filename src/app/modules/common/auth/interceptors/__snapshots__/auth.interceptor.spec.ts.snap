// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`insert auth token should insert token if signed in 1`] = `"Bearer test"`;
